File: go/internal/network/network_test.go
package network_test

import (
	"strings"
	"testing"

	"github.com/rootwarp/eth-utils/go/internal/network"
)

// TestConstants verifies the compile-time byte values.
func TestConstants(t *testing.T) {
	t.Run("DomainDeposit", func(t *testing.T) {
		want := [4]byte{0x03, 0x00, 0x00, 0x00}
		if network.DomainDeposit != want {
			t.Errorf("DomainDeposit = %v, want %v", network.DomainDeposit, want)
		}
	})

	t.Run("ZeroGenesisValidatorsRoot", func(t *testing.T) {
		var want [32]byte // all zeros
		if network.ZeroGenesisValidatorsRoot != want {
			t.Errorf("ZeroGenesisValidatorsRoot = %v, want all zeros", network.ZeroGenesisValidatorsRoot)
		}
	})
}

// TestLookupMainnet verifies mainnet fork bytes byte-for-byte.
func TestLookupMainnet(t *testing.T) {
	params, err := network.Lookup(network.Mainnet)
	if err != nil {
		t.Fatalf("Lookup(Mainnet) unexpected error: %v", err)
	}
	want := [4]byte{0x00, 0x00, 0x00, 0x00}
	if params.GenesisForkVersion != want {
		t.Errorf("GenesisForkVersion = %v, want %v", params.GenesisForkVersion, want)
	}
	if params.Name != network.Mainnet {
		t.Errorf("Name = %q, want %q", params.Name, network.Mainnet)
	}
}

// TestLookup verifies Lookup for all supported networks, new fields, and error cases.
func TestLookup(t *testing.T) {
	tests := []struct {
		net                    network.Network
		wantForkVersion        [4]byte
		wantChainID            uint64
		wantDepositContractHex string
	}{
		{
			net:                    network.Mainnet,
			wantForkVersion:        [4]byte{0x00, 0x00, 0x00, 0x00},
			wantChainID:            1,
			wantDepositContractHex: "0x00000000219ab540356cbb839cbe05303d7705fa",
		},
		{
			net:                    network.Hoodi,
			wantForkVersion:        [4]byte{0x10, 0x00, 0x09, 0x10},
			wantChainID:            560048,
			wantDepositContractHex: "0x00000000219ab540356cbb839cbe05303d7705fa",
		},
		{
			net:                    network.Sepolia,
			wantForkVersion:        [4]byte{0x90, 0x00, 0x00, 0x69},
			wantChainID:            11155111,
			wantDepositContractHex: "0x7f02c3e3c98b133055b8b348b2ac625669ed295d",
		},
		{
			net:                    network.Holesky,
			wantForkVersion:        [4]byte{0x01, 0x01, 0x70, 0x00},
			wantChainID:            17000,
			wantDepositContractHex: "0x4242424242424242424242424242424242424242",
		},
	}

	for _, tc := range tests {
		tc := tc
		t.Run(string(tc.net), func(t *testing.T) {
			params, err := network.Lookup(tc.net)
			if err != nil {
				t.Fatalf("Lookup(%q) error = %v, want nil", tc.net, err)
			}
			if params.Name != tc.net {
				t.Errorf("Name = %q, want %q", params.Name, tc.net)
			}
			if params.GenesisForkVersion != tc.wantForkVersion {
				t.Errorf("GenesisForkVersion = %v, want %v", params.GenesisForkVersion, tc.wantForkVersion)
			}
			if params.ChainID != tc.wantChainID {
				t.Errorf("ChainID = %d, want %d", params.ChainID, tc.wantChainID)
			}
			// DepositContractAddressHex returns lowercase 0x-prefixed.
			gotHex := strings.ToLower(params.DepositContractAddressHex())
			if gotHex != tc.wantDepositContractHex {
				t.Errorf("DepositContractAddressHex() = %q, want %q", gotHex, tc.wantDepositContractHex)
			}
		})
	}

	t.Run("Unknown_network_returns_descriptive_error", func(t *testing.T) {
		unknown := network.Network("goerli")
		_, err := network.Lookup(unknown)
		if err == nil {
			t.Fatal("Lookup(unknown) error = nil, want error")
		}
		if !strings.Contains(err.Error(), "goerli") {
			t.Errorf("error %q does not mention the unknown network name", err.Error())
		}
	})
}

// TestParseFlag verifies case-sensitive exact matching.
func TestParseFlag(t *testing.T) {
	validCases := []struct {
		input string
		want  network.Network
	}{
		{"mainnet", network.Mainnet},
		{"hoodi", network.Hoodi},
		{"sepolia", network.Sepolia},
		{"holesky", network.Holesky},
	}

	for _, tc := range validCases {
		tc := tc
		t.Run("valid_"+tc.input, func(t *testing.T) {
			got, err := network.ParseFlag(tc.input)
			if err != nil {
				t.Fatalf("ParseFlag(%q) error = %v, want nil", tc.input, err)
			}
			if got != tc.want {
				t.Errorf("ParseFlag(%q) = %q, want %q", tc.input, got, tc.want)
			}
		})
	}

	invalidCases := []string{
		"",
		"HOODI",
		"mainnet ",
		"goerli",
		"Mainnet",
		" mainnet",
		"MAINNET",
		"SEPOLIA",
		"Holesky",
	}

	for _, input := range invalidCases {
		input := input
		t.Run("invalid_"+input, func(t *testing.T) {
			_, err := network.ParseFlag(input)
			if err == nil {
				t.Errorf("ParseFlag(%q) error = nil, want error", input)
			}
		})
	}
}

// TestLookupByChainID verifies reverse lookup by chain ID for all 4 networks
// and an error case for an unknown chain ID.
func TestLookupByChainID(t *testing.T) {
	cases := []struct {
		chainID     uint64
		wantNetwork network.Network
		wantURL     string
	}{
		{1, network.Mainnet, "https://etherscan.io"},
		{560048, network.Hoodi, "https://hoodi.etherscan.io"},
		{11155111, network.Sepolia, "https://sepolia.etherscan.io"},
		{17000, network.Holesky, "https://holesky.etherscan.io"},
	}

	for _, tc := range cases {
		tc := tc
		t.Run(string(tc.wantNetwork), func(t *testing.T) {
			p, err := network.LookupByChainID(tc.chainID)
			if err != nil {
				t.Fatalf("LookupByChainID(%d) error = %v, want nil", tc.chainID, err)
			}
			if p.Name != tc.wantNetwork {
				t.Errorf("Name = %q, want %q", p.Name, tc.wantNetwork)
			}
			if p.ChainID != tc.chainID {
				t.Errorf("ChainID = %d, want %d", p.ChainID, tc.chainID)
			}
			if p.ExplorerURL != tc.wantURL {
				t.Errorf("ExplorerURL = %q, want %q", p.ExplorerURL, tc.wantURL)
			}
		})
	}

	t.Run("unknown_chain_ID", func(t *testing.T) {
		_, err := network.LookupByChainID(99999)
		if err == nil {
			t.Error("LookupByChainID(99999) error = nil, want error")
		}
	})
}

// TestLookupExplorerURL verifies ExplorerURL is set for all known networks.
func TestLookupExplorerURL(t *testing.T) {
	for _, n := range []network.Network{network.Mainnet, network.Hoodi, network.Sepolia, network.Holesky} {
		n := n
		t.Run(string(n), func(t *testing.T) {
			p, err := network.Lookup(n)
			if err != nil {
				t.Fatalf("Lookup(%q) error = %v", n, err)
			}
			if p.ExplorerURL == "" {
				t.Errorf("ExplorerURL is empty for network %q", n)
			}
			if !strings.HasPrefix(p.ExplorerURL, "https://") {
				t.Errorf("ExplorerURL = %q: expected https:// prefix", p.ExplorerURL)
			}
		})
	}
}

// TestDepositContractAddressHex verifies the hex formatting helper.
func TestDepositContractAddressHex(t *testing.T) {
	params, err := network.Lookup(network.Holesky)
	if err != nil {
		t.Fatalf("Lookup(Holesky) error = %v", err)
	}
	got := params.DepositContractAddressHex()
	if !strings.HasPrefix(got, "0x") {
		t.Errorf("DepositContractAddressHex() = %q: missing 0x prefix", got)
	}
	// hex body is 40 chars (20 bytes)
	if len(got) != 42 {
		t.Errorf("DepositContractAddressHex() len = %d, want 42", len(got))
	}
}

File: rust/crates/core/tests/network_test.rs
//! Ported from go/internal/network/network_test.go.
//!
//! Black-box tests against the public network surface. Two deliberate
//! adaptations to the Rust API:
//!   * Go's `network.Lookup(net) (Params, error)` is infallible in Rust
//!     (`network::lookup(Network) -> Params`) because `Network` is a closed
//!     enum with no "unknown" value. The Go "unknown network" subtest is
//!     therefore ported against `lookup_name`, which is the read-side entry
//!     that accepts an arbitrary string.
//!   * `network.Network("goerli")` cannot be constructed in Rust, so string
//!     inputs go through `lookup_name` / `parse_flag`.

use eth_deposit_core::network::{
    self, lookup, lookup_by_chain_id, lookup_name, parse_flag, Network, DOMAIN_DEPOSIT,
    ZERO_GENESIS_VALIDATORS_ROOT,
};

// Go: TestConstants/DomainDeposit
#[test]
fn constants_domain_deposit() {
    assert_eq!(DOMAIN_DEPOSIT, [0x03, 0x00, 0x00, 0x00]);
}

// Go: TestConstants/ZeroGenesisValidatorsRoot
#[test]
fn constants_zero_genesis_validators_root() {
    assert_eq!(ZERO_GENESIS_VALIDATORS_ROOT, [0u8; 32]);
}

// Go: TestLookupMainnet
#[test]
fn lookup_mainnet() {
    let params = lookup(Network::Mainnet);
    assert_eq!(params.genesis_fork_version, [0x00, 0x00, 0x00, 0x00]);
    assert_eq!(params.name, Network::Mainnet);
}

// Go: TestLookup (per-network field table).
#[test]
fn lookup_all_networks() {
    struct Case {
        net: Network,
        fork_version: [u8; 4],
        chain_id: u64,
        deposit_contract_hex: &'static str,
    }
    let cases = [
        Case {
            net: Network::Mainnet,
            fork_version: [0x00, 0x00, 0x00, 0x00],
            chain_id: 1,
            deposit_contract_hex: "0x00000000219ab540356cbb839cbe05303d7705fa",
        },
        Case {
            net: Network::Hoodi,
            fork_version: [0x10, 0x00, 0x09, 0x10],
            chain_id: 560048,
            deposit_contract_hex: "0x00000000219ab540356cbb839cbe05303d7705fa",
        },
        Case {
            net: Network::Sepolia,
            fork_version: [0x90, 0x00, 0x00, 0x69],
            chain_id: 11155111,
            deposit_contract_hex: "0x7f02c3e3c98b133055b8b348b2ac625669ed295d",
        },
        Case {
            net: Network::Holesky,
            fork_version: [0x01, 0x01, 0x70, 0x00],
            chain_id: 17000,
            deposit_contract_hex: "0x4242424242424242424242424242424242424242",
        },
    ];

    for c in &cases {
        let params = lookup(c.net);
        assert_eq!(params.name, c.net, "name for {:?}", c.net);
        assert_eq!(
            params.genesis_fork_version, c.fork_version,
            "fork version for {:?}",
            c.net
        );
        assert_eq!(params.chain_id, c.chain_id, "chain id for {:?}", c.net);
        assert_eq!(
            params.deposit_contract_address_hex().to_lowercase(),
            c.deposit_contract_hex,
            "deposit contract hex for {:?}",
            c.net
        );
    }
}

// Go: TestLookup/Unknown_network_returns_descriptive_error
// Adapted to `lookup_name`, the string-keyed read-side lookup (see module doc).
#[test]
fn lookup_name_unknown_returns_descriptive_error() {
    let err = lookup_name("goerli").unwrap_err();
    assert!(
        err.to_string().contains("goerli"),
        "error {err:?} does not mention the unknown network name"
    );
}

// Go: TestParseFlag (valid + invalid case-sensitive matching).
#[test]
fn parse_flag_valid() {
    let cases = [
        ("mainnet", Network::Mainnet),
        ("hoodi", Network::Hoodi),
        ("sepolia", Network::Sepolia),
        ("holesky", Network::Holesky),
    ];
    for (input, want) in cases {
        let got =
            parse_flag(input).unwrap_or_else(|e| panic!("parse_flag({input:?}) error: {e:?}"));
        assert_eq!(got, want, "parse_flag({input:?})");
    }
}

// Go: TestParseFlag (invalid cases — all must error).
#[test]
fn parse_flag_invalid() {
    let invalid = [
        "", "HOODI", "mainnet ", "goerli", "Mainnet", " mainnet", "MAINNET", "SEPOLIA", "Holesky",
    ];
    for input in invalid {
        assert!(
            parse_flag(input).is_err(),
            "parse_flag({input:?}) should have errored"
        );
    }
}

// Go: TestLookupByChainID (reverse lookup + explorer URL).
#[test]
fn lookup_by_chain_id_all() {
    let cases = [
        (1u64, Network::Mainnet, "https://etherscan.io"),
        (560048, Network::Hoodi, "https://hoodi.etherscan.io"),
        (11155111, Network::Sepolia, "https://sepolia.etherscan.io"),
        (17000, Network::Holesky, "https://holesky.etherscan.io"),
    ];
    for (chain_id, want_network, want_url) in cases {
        let p = lookup_by_chain_id(chain_id)
            .unwrap_or_else(|e| panic!("lookup_by_chain_id({chain_id}) error: {e:?}"));
        assert_eq!(p.name, want_network, "network for chain {chain_id}");
        assert_eq!(p.chain_id, chain_id, "chain id round-trip");
        assert_eq!(
            p.explorer_url, want_url,
            "explorer url for chain {chain_id}"
        );
    }
}

// Go: TestLookupByChainID/unknown_chain_ID
#[test]
fn lookup_by_chain_id_unknown() {
    assert!(lookup_by_chain_id(99999).is_err());
}

// Go: TestLookupExplorerURL
#[test]
fn lookup_explorer_url_present() {
    for n in Network::ALL {
        let p = lookup(n);
        assert!(!p.explorer_url.is_empty(), "explorer url empty for {n:?}");
        assert!(
            p.explorer_url.starts_with("https://"),
            "explorer url for {n:?} lacks https:// prefix: {}",
            p.explorer_url
        );
    }
}

// Go: TestDepositContractAddressHex
#[test]
fn deposit_contract_address_hex_format() {
    let params = lookup(Network::Holesky);
    let got = params.deposit_contract_address_hex();
    assert!(got.starts_with("0x"), "missing 0x prefix: {got}");
    // 0x + 40 hex chars (20 bytes).
    assert_eq!(got.len(), 42, "deposit contract hex length");
}

// Parity guard for the `network` module path re-export used by callers.
#[test]
fn lookup_name_known() {
    let p = network::lookup_name("mainnet").expect("mainnet is known");
    assert_eq!(p.name, Network::Mainnet);
}
